torchtext/data/iterator.py

```diff
@@ -1,11 +1,34 @@
 import math
 import random
-
-import torch
+from contextlib import contextmanager
+from copy import deepcopy
 
 from .batch import Batch
 from .dataset import Dataset
 
+class RandomShuffler(object):
+    """Use random functions while keeping track of the random state to make it
+    reproducible and deterministic."""
+
+    def __init__(self, random_state=None):
+        self.random_state = random_state
+        if self.random_state is None:
+            self.random_state = random.getstate()
+
+    @contextmanager
+    def use_internal_state(self):
+        """Use a specific RNG state."""
+        old_state = random.getstate()
+        random.setstate(self.random_state)
+        yield
+        self.random_state = random.getstate()
+        random.setstate(old_state)
+
+    def __call__(self, data):
+        """Shuffle and return a new list."""
+        with self.use_internal_state():
+            list_data = list(data)
```

Comment:
`len(data)` doesn't work if `data` is an iterator, which is the case in the `pool` function when calling `random_shuffler(batch(...))`. In that case, should I convert to list outside of the `RandomShuffler` class?